### ISSUE TITLE
All assemblies marked as run even when selecting items from a single assembly

### ISSUE THREAD

**Issue by CharliePoole** [TestCentric/testcentric-gui#431]
Load more than one assembly using the command-line. Select a fixture in one assembly and run the tests. Note that the assemblies __not__ containing the fixture display with a green icon, as if they had been run successfully.

Load a `.nunit` project containing multiple assemblies. Follow the same process. Again, the assemblies __not__ containing the fixture are displayed as if they were run.

It's likely that this is due to the framework, which is being told to run a test using the filter that was constructed. The engine should either ignore this notice or - even better - not pass the run command to any assembly that contains no selected tests.

This is a low-priority bug and will be removed from the 1.1 milestone if it proves complicated.

**Comment by CharliePoole** [comment]
This is much more complicated than it seemed. I'm removing it from the 1.1 milestone.

**Comment by CharliePoole** [comment]
Recording info as to why this is so complicated, to assist in designing a solution.

### TL;DR

It seems as if none of the three architectural layers alone can handle this. Any solution will probably require fixes in two layers or possibly all three.

### Details

Let's say an assembly has no tests. There are two possibilities...

1. It does not reference a known test framework.
2. It references a test framework, but simply contains no identified tests.

The first case is already taken care of... it results in an error unless the we tell the run we want to skip assemblies with no tests. This happens, for example, when running a VS solution containing both test and non-test assemblies.

It's the second case that troubles us here. This could be dealt with in any of the three layers of the NUnit architecture: framework, engine or runner. Let's look at each one.

### Framework

The framework is told to run an assembly, which turns out to have no tests. This could mean a number of different things:

 * It's an error, there are supposed to be tests
 * A filter was used across a single assembly and nothing was found.
 * A filter was used across multiple assemblies and nothing was found in __this__ assembly.
 * Tests are generated dynamically and none were generated on this occasion.
 * Tests are added to this assembly manually as needed. None are currently needed.

When the framework is invoked to run tests, most of the above information is not available, at least currently. The framework has a filter, which is either empty or not. It can try to run tests and return either success, a warning or failure when none are found. Currently, it returns success.

It could return a special code, i.e. "No tests found", but that would not provide any more information than is now made available in the result that is returned.

Alternatively (I'm looking ahead here) some higher level code could pass in something that tells it what to do if no tests are found. Or, the higher level code might avoid running the assembly at all if it "knows" that know tests will be found. This could be the case in the GUI at least in some cases.

### Engine

The same cases apply to the framework driver when invoking the framework. The same information is available to the driver as the framework - and the same lack of information as well! Anything done in the framework could probably happen in the driver with the added benefit that the framework would not need to be invoked.

Higher levels of the engine could do the same thing. They could also analyze the filter in order to determine whether only a certain assembly or assemblies need to be run. For example, if all filters are ID filters, the filter prefixes would indicate the assemblies to run. Or the engine could cache a list of categories present in each assembly.

The main drawback of some of these approaches in the engine is that it requires analysis to determine things that could be more easily determined at the runner level.

### Runner

In this issue, we are talking about the GUI runner, but any solution we come up with would need to be applied by other runners as well. Still...

In the reported issue, the user selected a number of tests in the tree, all of them under the same assembly. It's clear to the runner that only that assembly should be run. However, due to the way the architecture works currently, it's necessary to __try__ to run those tests in every assembly. That's because test filters are created in the runner and passed __through__ the engine to the framework. There's no way for the runner to tell the engine not to bother with other assemblies.

In the absence of such a setting, the GUI could record the fact that it only cares about certain assemblies and suppress events it receives from other assemblies. We would have to think about any side effects of doing so.

**Comment by CharliePoole** [comment]
@nunit/framework-and-engine-team Although this is a TestCentric GUI issue, it clearly has impact on NUnit unless a runner-only solution is used. Comments appreciated!

**Comment by CharliePoole** [comment]
This issue has been without action for too long! As a simplification, I'll look at a solution in the runner itself. We actually need a non-framework solution since we'd like the fix to work for older versions of the framework anyway. It may also involve changes to the __TestCentric__ engine, which is only used by the GUI anyway.
